api/v1alpha1/api_key_auth_types.go

```diff
@@ -0,0 +1,69 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+const APIKeysSecretKey = "credentials"
+
+// APIKeyAuth defines the configuration for the API Key Authentication.
+type APIKeyAuth struct {
+	// Credentials is the Kubernetes secret which contains the API keys.
+	// This is an Opaque secret.
+	// Each API key is stored in the key representing the client id,
+	// which can be used in AllowedClients to authorize the client in a simple way.
+	// If this is attached to a HTTPRoute while the other SecurityPolicies with APIKeyAuth is attached to a Gateway,
+	// this credentials takes precedence.
+	// This field is optional, but it doesn't make sense that any attached APIKeyAuth for the route/gateway doesn't have credentials at all.
+	//
+	// +optional
+	Credentials gwapiv1.SecretObjectReference `json:"credentials,omitempty"`
+
+	// KeySources is where to fetch the key from the coming request.
+	// The value from the first source that has a key will be used.
+	// If this is attached to a HTTPRoute while the other SecurityPolicies with APIKeyAuth is attached to a Gateway,
+	// this key sources takes precedence.
+	// This field is optional, but it doesn't make sense that any attached APIKeyAuth for the route/gateway doesn't have key sources at all.
+	//
+	// +optional
+	KeySources []*KeySource `json:"keySources,omitempty"`
+
+	// AllowedClients is a list of clients that are allowed to access the route or vhost.
+	// The clients listed here should be subset of the clients listed in the `Credentials` to provide authorization control
+	// after the authentication is successful. If the list is empty, then all authenticated clients
+	// are allowed. This provides very limited but simple authorization.
+	//
+	// +optional
+	AllowedClients []string `json:"allowedClients,omitempty"`
```

Comment:
Are there any use case where an API key is authenticated but not allowed to access the targeted Route/Gateway?

Reply:
The use case is:
1. the gateway has API Key Auth configured globally. e.g., `Credentials` has `admin: key1` and `user: key2`.
2. the route also has per-route API Key Auth, and you can define this route is only for `admin` with `AllowedClients`.

Reply:
In such case, this route should only has admin: key1 credential attached.

Reply:
I see, I missed that point. 
But, I'm not sure if it's the best interface for users that have to prepare secret per access group. Also, the same secret for each client would be spread across different secrets.

So, I still see the benefit of AllowedClients there; users can have only one secret as a source of credentials and they can select which key(s) is enabled on certain routes.